README
sd2iec - a controller/interface adapting storage devices to the CBM serial bus
Copyright (C) 2007-2014  Ingo Korb <ingo@akana.de>
Parts based on code from others, see comments in main.c for details.
JiffyDos send based on code by M.Kiesel
Fat LFN support and lots of other ideas+code by Jim Brain
Final Cartridge III fastloader support by Thomas Giesel
IEEE488 support by Nils Eilers

Free software under GPL version 2 ONLY, see comments in main.c and
COPYING for details.

FIXME:This file still needs to be expanded. A lot.
FIXME: sprinkle mentions of IEEE488 where appropiate

Deprecation notices
===================
The following feature(s) will be removed in the next release:
- M2I support
   M2I support has been redundant since the introduction of transparent
   P00 support. To continue to use your M2I-format software, convert
   your files to P00 format (e.g. with m2itopc64.c) and set your device
   to extension mode 2 (XE2).

Introduction:
=============
sd2iec is firmware, used in hardware designs like MMC2IEC, SD2IEC, or uIEC,
that allows the Commodore serial bus to access removable storage devices
(MMC, SD, CF) - think of it as a 1541 with a modern storage medium instead
of disks. The project was inspired by (and uses a few bits of code from)
MMC2IEC[1] by Lars Pontoppidan and once ran on the same hardware before it
grew too big for the ATmega32 used there.

Currently, the firmware provide good DOS and file-level compatibility with CBM
drives, but much work remains.  Unless specifically noted, anything that tries
to execute code on the 1541 will not work, this includes every software
fastloader.

[1] Homepage: http://pontoppidan.info/lars/index.php?proj=mmc2iec

Please note: Whenever this file talks about "D64 images" the text applies to
all Dxx image types, i.e. D64/D71/D81/DNP unless specifically noted.

If you are the author of a program that needs to detect sd2iec for
some reason, DO NOT use M-R for this purpose. Use the UI command
instead and check the message you get for "sd2iec" and "uiec" instead.


Supported commands:
===================
- General notes:
  Any command not listed below is currently not supported.

- Directory filters:
  To show only directories, both =B (CMD-compatible) and =D can be used.
  On a real Commodore drive D matches everything.
  To include hidden files in the directory, use *=H - on a 1541 this doesn't
  do anything. sd2iec marks hidden files with an H after the lock mark,
  i.e. "PRG<H" or "PRG H".

  CMD-style "short" and "long" directory listings with timestamps are supported
  ("$=T"), including timestamp filters. Please read a CMD manual for the syntax
  until this file is updated.

- Partition directory:
  The CMD-style partition directory ($=P) is supported, including filters
  ($=P:S*). All partitions are listed with type "FAT", although this could
  change to "NAT" later for compatibility.

- CD/MD/RD:
  Subdirectory access is compatible to the syntax used by the CMD drives,
  although drive/partition numbers are completely ignored.

  Quick syntax overview:
    CD:_         changes into the parent dir (_ is the left arrow on the C64)
    CD_          dito
    CD:foo       changes into foo
    CD/foo       dito
    CD//foo      changes into \foo
    CD/foo/:bar  changes into foo\bar
    CD/foo/bar   dito

  You can use wildcards anywhere in the path. To change into an M2I or D64
  image the image file must be the last component in the path, either
  after a slash or a colon character.

  MD uses a syntax similiar to CD and will create the directory listed
  after the colon (:) relative to any directory listed before it.

    MD/foo/:bar  creates bar in foo
    MD//foo/:bar creates bar in \foo

  RD can only remove subdirectories of the current directory.

    RD:foo       deletes foo

  CD is also used to mount/unmount image files. Just change into them
  as if they were a directory and use CD:_ (left arrow on the C64) to leave.
  Please note that image files are detected by file extension and file size
  and there is no reliable way to see if a file is a valid image file.

- CP, C<Shift-P>
  This changes the current partition, see "Partitions" below for details.

- C:
  File copy command, should be CMD compatible. The syntax is
   C[partition][path]:targetname=[[partition][path]:]sourcename[,[[p][p]:]sourcename...]
  You can use this command to copy multiple files into a single target
  file in which case all source files will be appended into the target
  file. Parsing restarts for every source file name which means that
  every source name is assumed to be relative to the current directory.
  You can use wildcards in the source names, but only the first
  file matching will be copied.

  Copying REL files should work, but isn't tested well. Mixing REL and
  non-REL files in an append operation isn't supported.

- D
  Direct sector access, this is a command group introduced by sd2iec.
  Some Commodore drives use D for disk duplication between two drives
  in the same unit, an attempt to use that command with sd2iec should
  result in an error message.

  D has three subcommands: DI (Info), DR (Read) and DW (Write).
  Each of those commands requires a buffer to be opened (similiar
  to U1/U2), but due to the larger sector size of the storage devices
  used by sd2iec it needs to be a large buffer of size 2 (512 bytes)
  or larger. The exception is the DI command with page set to 0,
  its result will always fir into a standard 256 byte buffer.
  If you try to use one of the commands with a buffer that is too
  small a new error message is returned, "78,BUFFER TOO SMALL,00,00".

  In the following paragraphs the secondary address that was used
  to open the buffer is called "bufchan".

  - DI
    In BASIC notation the command format is
      "DI"+chr$(bufchan)+chr$(device)+chr$(page)

    "device" is the number of the physical device to be queried,
    "page" the information page to be retrieved. Currently the
    only page implemented is page 0 which will return the
    following data structure:
     1 byte : Number of valid bytes in this structure
              This includes this byte and is meant to provide
              backwards compatibility if this structure is extended
              at a later time. New fields will always be added to
              the end so old programs can still read the fields
              they know about.
     1 byte : Highest diskinfo page supported
              Always 0 for now, will increase if more information
              pages are added (planned: Complete ATA IDENTIFY
              output for IDE and CSD for SD)
     1 byte : Disk type
              This field identifies the device type, currently
              implemented values are:
                0  IDE
                2  SD
                3  (reserved)
     1 byte : Sector size divided by 256
              This field holds the sector size of the storage device
              divided by 256.
     4 bytes: Number of sectors on the device 
              A little-endian (same byte order as the 6502) value
              of the number of sectors on the storage device.
              If there is ever a need to increase the reported
              capacity beyond 2TB (for 512 byte sectors) this
              field will return 0 and a 64-bit value will be added
              to this diskinfo page.

    If you want to determine if there is a device that responds
    to a given number, read info page 0 for it. If there is no
    device present that corresponds to the number you will see
    a DRIVE NOT READY error on the error channel and the
    "number of valid bytes" entry in the structure will be 0.

    Do not assume that device numbers are stable between releases
    and do not assume that they are continuous either. To scan
    for all present devices you should query at least 0-7 for now,
    but this limit may increase in later releases.

  - DR/DW
    In BASIC notation the command format would be
      "DR"+chr$(bufchan)+chr$(device)
          +chr$(sector AND 255)
          +chr$((sector/256) AND 255)
          +chr$((sector/65536) AND 255)
          +chr$((sector/16777216) AND 255)
    (or "DW" instead of "DR)
    but this won't work on the C64 because AND does not accept
    parameters larger than 32767. The principle should be clear
    though, the last four bytes are a 32 bit sector number in
    little-endian byte order.

    DR reads the sector to the buffer, DW writes the contents
    of the buffer to the sector. Both commands will update the
    error channel if an error occurs, for DR the 20,READ ERROR
    was chosen to represent read errors; for write problems
    during DW it sets 25,WRITE ERROR for errors and 26,WRITE
    PROTECT ON if the device is read-only.

- G-P
  Get partition info, see CMD FD/HD manual for details. The reported
  information is partially faked, feedback is welcome.

- P
  Positioning doesn't just work for REL files but also for regular
  files on a FAT partition. When used for regular files the format
  is "P"+chr$(channel)+chr$(lo)+chr$(midlo)+chr$(midhi)+chr$(hi)
  which will seek to the 0-based offset hi*2^24+midhi*65536+256*midlo+lo
  in the file. If you send less than four bytes for the offset, the
  missing bytes are assumed to be zero.

- N:
  Format works only if a D64 image is already mounted. This command will
  be ignored for DNP images unless the current directory is the root
  directory of the DNP image.

- R
  Renaming files should work the same as it does on CMD drives, although
  the errors flagged for invalid characters in the name may differ.

- S:
  Name matching is fully supported, directories are ignored.
  Scratching of multiple files separated by , is also supported with no
  limit to the number of files except for the maximum command line length
  (usually 100 to 120 characters).

- T-R and T-W
  If your hardware features RTC support the commands T-R (time read) and T-W
  (time write) are available. If the RTC isn't present, both commands return
  30,SYNTAX ERROR,00,00; if the RTC is present but not set correctly T-R will
  return 31,SYNTAX ERROR,00,00.

  Both commands expect a fourth character that specifies the time format
  to be used. T-W expects that the new time follows that character
  with no space or other characters inbetween. For the A, B and D
  formats, the expected input format is exactly the same as returned
  by T-R with the same format character; for the I format the day of
  week is ignored and calculated based on the date instead.

  The possible formats are:
   - "A"SCII: "SUN. 01/20/08 01:23:45 PM"+CHR$(13)
      The day-of-week string can be any of "SUN.", "MON.", "TUES", "WED.",
      "THUR", "FRI.", "SAT.". The year field is modulo 100.

   - "B"CD or "D"ecimal:
     Both these formats use 9 bytes to specify the time. For BCD everything
     is BCD-encoded, for Decimal the numbers are sent/parsed as-is.
      Byte 0: Day of the week (0 for sunday)
           1: Year (modulo 100 for BCD; -1900 for Decimal, i.e. 108 for 2008)
           2: Month (1-based)
           3: Day (1-based)
           4: Hour   (1-12)
           5: Minute (0-59)
           6: Second (0-59)
           7: AM/PM-Flag (0 is AM, everything else is PM)
           8: CHR$(13)

      When the time is set a year less than 80 is interpreted as 20xx.

   - "I"SO 8601 subset: "2008-01-20T13:23:45 SUN"+CHR$(13)
     This format complies with ISO 8601 and adds a day of week
     abbreviation using the same table as the A format, but omitting
     the fourth character. When it is used with T-W, anything beyond
     the seconds field is ignored and the day of week is calculated
     based on the specified date. The year must always be specified
     including the century if this format is used to set the time.
     To save space, sd2iec only accepts this particular date/time
     representation when setting the time with T-WI and no other ISo
     8601-compliant representation.

- U0
  Device address changing with "U0>"+chr$(new address) is supported,
  other U0 commands are currently not implemented.

- U1/U2/B-R/B-W
  Block reading and writing is fully supported while a D64 image is mounted.

- B-P
  Supported, not checked against the original rom at all.

- UI+/UI-
  Switching the slightly faster bus protocol for the VC20 on and off works,
  it hasn't been tested much though.

- UI/UJ
  Soft/Hard reset - UI just sets the "73,..." message on the error channel,
  UJ closes all active buffers but doesn't reset the current directory,

NEWS
201x-xx-xx - release 1.0.0
        - Bugfix: use valid default date if RTC is not detected
        - formatting for D71/D81/DNP images
        - DolphinDOS parallel speeder support
        - PREV button mounts last line of autoswap.lst
        - removed XB command
        - explicit PETSCII marker in swaplists
        - load previous file when "*" is used
        - redirect C128 boot sector to bootsect.128 on FAT
        - ISO8601-like format for T-R and T-W
        - autoswap list generation
        - persistent mini-filesystem in EEPROM
        - Maniac Mansion + Zak McKracken loaders

2012-02-26 - release 0.10.3
        - Bugfix: Un-break I2C display communication

2011-12-18 - release 0.10.2
        - Bugfix: End of generated raw directory was incorrect
        - Bugfix: Various compilation fixes when fastloaders are disabled
        - ULoad M3 automatically exits when ATN becomes active
        - Minor code space improvements
        - New fastloader: ELoad Version 1 (EasyProg)

2010-12-25 - release 0.10.1
        - Bugfix: Remove remains of previous commands while parsing B-*
        - Bugfix: Fix response of UI command for unknown suffix chars
        - small improvement of JiffyDOS transfer speeds

2010-11-07 - release 0.10.0
        - Deprecation notice:
          E-R and E-W commands will be removed in the next release
        - Improved reliability of Nippon loader

2010-10-31 - release 0.10.0pre5
        - Bugfix: Avoid duplicate file name when DEL/DIR matches
        - Bugfix: Reduce chance of data corruption in Wheels
        - Bugfix: Slightly relax IEC timing while waiting for command
        - Added AR6 1581 fastloader+saver (PAL)

2010-10-26 - release 0.10.0pre4
        - Bugfix: Wheels with SuperCPU timing issue corrected
        - Bugfix: Rare problem with ~ in FAT file names fixed
        - Added compile-time option to capture fastloader code
        - Added Wheels 4.4 fast loaders
        - Added another variation of the Epyx Fastload cart
        - Added Nippon fast loader

2010-10-16 - release 0.10.0pre3
        - Bugfix: Wheels 128 loader detection fixed
        - Added a second Epyx Fastload variation

2010-10-15 - release 0.10.0pre2
        - Bugfix: G-P answer fixed
        - Bugfix: Strip 0x0d from file names when appropiate
        - Wheels support

2010-08-06 - release 0.10.0pre1
        - GEOS 2.0 support (d64/d71/d81 images only)
        - File-based M-R emulation
        - Removed "oldnkc" configuration

2010-08-02 - release 0.9.3
        - Bugfix: Extension hiding works again with names >12 characters
        - Bugfix: Increase data hold time during standard TALK
        - Testing aid: card detect-to-LED function

2010-06-21 - release 0.9.2
        - Bugfix: Rename was broken on FAT
        - Turbodisk is now disabled when running on internal RC clock
        - button debounce timeout increased

2010-03-07 - release 0.9.1
        - Bugfix: Filling unused REL records used wrong offsets
        - Bugfix: Unbreak case-insensitive matching of FAT file names
        - Bugfix: Fix raw directory mode for D64/D71/D81
        - Bugfix: Tell the display about REL files too
        - add preliminary support for the ATmega1284P

2010-02-28 - release 0.9.0
        - Bugfix: REL offset calculation for records 256,512,...
        - Bugfix: Non-REL P command parameters now match README
        - Bugfix: Reading after EOI behaves correctly now
        - Red LED only active while unwritten data in memory
        - Remove XC,XJ commands
        - New fastloader: ULoad Model 3
        - New fastloader: G.I. Joe (said to be commonly ripped)
        - New fastloader: Epyx FastLoad cartridge
        - Support reading raw directory ($ with secondary > 0)
        - "display disk images as directories"-mode
        - DNP support
        - remote display support
        - build system doesn't require gawk anymore

2009-12-16 - release 0.8.3
        - Bugfix: Really prevent random corruption on FAT32 media
        - Bugfix: Free buffer after copying files

2009-09-27 - release 0.8.2
        - Bugfix: Prevent random storage corruption on FAT32 media
        - Bugfix: Image files not mountable if read-only file or media
        - Bugfix: Do not clear error channel if sec. 15 is closed
        - Add a dummy RTC read during init (to update year if required)
        - Enable I2C RTC support in LarsP configuration for Final Expansion 3
        - Acknowledge ATN during initialisation to stall C64 if required

2009-05-24 - release 0.8.1a
        - Fix bootloader version number generation

2009-05-20 - release 0.8.1
        - Fix uIEC/IDE SD card change detection
        - Dreamload support for uIEC
        - Mark the RTC time as valid after setting the clock
        - Delay SPI initialisation until absolutely required
        - Show disk change confirmation even when the error blink is active
        - Exit the FC3 fastloader if ATN is active
        - Fix indentation of 1000+ block files
        - Remove 0-byte from the long version message
        - Fix multiple save-with-replace bugs
        - Ignore directories when matching file name wildcards

2009-03-08 - release 0.8.0
        - direct sector access of the real storage device
        - ignore DEL files in file_open
        - drive mapper
        - EXOS V3 fastloader support
        - requires avr-libc 1.6.x

2008-10-02 - release 0.8.0pre1
        - Implement Dreamload emulation
        - RTC support
        - Dataflash support
        - Multi-file scratch
        - Copy command
        - Large buffer support
        - Final Cartridge 3 freezed file fastloader support
        - D71/D81 disk image support
        - parse FSINFO sector for faster FAT32 free space calculation
        - partial REL file support

2008-07-09 - release 0.7.3
        - Bugfix: Free buffers on error in M2I and D64
          (fixes "D64 size increases to 1.1M when its free space runs out")

2008-06-10 - release 0.7.2

src/ff.h
/*--------------------------------------------------------------------------/
/  FatFs - FAT file system module include file  R0.06        (C)ChaN, 2008
/---------------------------------------------------------------------------/
/ FatFs module is an experimenal project to implement FAT file system to
/ cheap microcontrollers. This is a free software and is opened for education,
/ research and development under license policy of following trems.
/
/  Copyright (C) 2008, ChaN, all right reserved.
/
/ * The FatFs module is a free software and there is no warranty.
/ * You can use, modify and/or redistribute it for personal, non-profit or
/ * commercial use without any restriction under your responsibility.
/ * Redistributions of source code must retain the above copyright notice.
/
/---------------------------------------------------------------------------*/

#ifndef _FATFS

#define _MCU_ENDIAN     1
/* The _MCU_ENDIAN defines which access method is used to the FAT structure.
/  1: Enable word access.
/  2: Disable word access and use byte-by-byte access instead.
/  When the architectural byte order of the MCU is big-endian and/or address
/  miss-aligned access results incorrect behavior, the _MCU_ENDIAN must be set
/  to 2. If it is not the case, it can be set to 1 for good code efficiency. */

#define _FS_READONLY    0
/* Setting _FS_READONLY to 1 defines read only configuration. This removes
/  writing functions, f_write, f_sync, f_unlink, f_mkdir, f_chmod, f_rename,
/  f_truncate and useless f_getfree. */

#define _FS_MINIMIZE    0
/* The _FS_MINIMIZE option defines minimization level to remove some functions.
/  0: Full function.
/  1: f_stat, f_getfree, f_unlink, f_mkdir, f_chmod, f_truncate and f_rename are removed.
/  2: f_opendir and f_readdir are removed in addition to level 1.
/  3: f_lseek is removed in addition to level 2. */

#define _USE_STRFUNC    0
/* To enable string functions, set _USE_STRFUNC to 1 or 2. */

#define _DRIVES     MAX_DRIVES
/* Number of physical drives to be used. This affects the size of internal
 * table. (when using _USE_DRIVE_PREFIX */

#define _USE_MKFS   0
/* When _USE_MKFS is set to 1 and _FS_READONLY is set to 0, f_mkfs function is
/  enabled. */

#define _MULTI_PARTITION    1
/* When _MULTI_PARTITION is set to 0, each logical drive is bound to the same
/  physical drive number and can mount only 1st primary partition.
/
/  When it is set to 1, the low _PARTITION_MASK bits of each partition represent
/  the partition and the high (8-_PARTITION_MASK) bits represent the physical
/  drive */

#define _PARTITION_MASK     4

#define _USE_FSINFO 1
/* To enable FSInfo support on FAT32 volume, set _USE_FSINFO to 1. */

#define _USE_SJIS   0
/* When _USE_SJIS is set to 1, Shift-JIS code transparency is enabled, otherwise
/  only US-ASCII(7bit) code can be accepted as file/directory name. */

#define _USE_NTFLAG 0
/* When _USE_NTFLAG is set to 1, upper/lower case of the file name is preserved.
/  Note that the files are always accessed in case insensitive. */

#define _USE_CHDIR 0

#define _USE_CURR_DIR 1

#define _USE_LFN 1

/* Maximum number of characters to return for a LFN */
/* The buffer used for FILINFO.lfn must be at least */
/* _MAX_LFN_LENGTH+1 characters long!               */
/* Note that if _USE_LFN_DBCS is set, this value    */
/* represents the characters needed, not bytes      */
#define _MAX_LFN_LENGTH 20

/* When _USE_LFN_DBCS is set to 1, FILINFO.lfn will contain a DBCS string, not
/  a simple ASCII string  */
#define _USE_LFN_DBCS 0

/* When set to 1, All FIL objects will use the buffer.  This reduces memory
/  requirements as open files will only require space for a FIL object, but
/  operate slower. When set, ff.c will behave like tff.c, but will allow
/  multiple filesystems.  */
#define _USE_FS_BUF 1

/* When set to 1, All objects will use a static  buffer.  This reduces memory
/  requirements to the absolute minimum ~512 bytes for the buffer, but will
/  operate slower.  This option can only be set if _USE_FS_BUF is set.  */
#define _USE_1_BUF 1

/* If set to 1, FatFs will manage the FATFS structures after mounting.  If
/  set to 0, the caller must send the correct drive FATFS structure for each
/  call.  Normally, this should be set to 1, but if the caller wants to use
/  low level l_* functions or skip sending the drive number in the path string,
/  this must be turned off.  */
#define _USE_DRIVE_PREFIX 0

/* If set to 1, FatFS will delay mounting the drive until first use.  Normally,
/  this should be turned on.  However, it cannot be used with
/  _USE_DRIVE_PREFIX = 0  */
#define _USE_DEFERRED_MOUNT 0

/* New features in 0.05a, not required yet */
#define _USE_TRUNCATE 0
#define _USE_UTIME   0

#include "integer.h"

#if _USE_LFN_DBCS != 0
#define S_LFN_OFFSET 26
#define S_LFN_INCREMENT 2
#else
#define S_LFN_OFFSET 13
#define S_LFN_INCREMENT 1
#endif

/* Definitions corresponds to multiple sector size (not tested) */
#define S_MAX_SIZ   512U            /* Do not change */
#if S_MAX_SIZ > 512U
#define SS(fs)  ((fs)->s_size)
#else
#define SS(fs)  512U
#endif

#if _USE_1_BUF == 1 && _USE_FS_BUF == 0
#error You can only use 1_BUF with _USE_FS_BUF at present
#define _USE_1_BUF 0
#endif

typedef struct _BUF {
  DWORD sect;
  BYTE  dirty;              /* dirty flag (1:must be written back) */
//BYTE  pad1;
#if _USE_1_BUF != 0
  struct _FATFS *fs;
#endif
  BYTE  data[S_MAX_SIZ];    /* Disk access window for Directory/FAT */
} BUF;

/* File system object structure */
typedef struct _FATFS {
  //WORD    id;             /* File system mount ID */
    WORD    n_rootdir;      /* Number of root directory entries */
    DWORD   sects_fat;      /* Sectors per fat */
    DWORD   max_clust;      /* Maximum cluster# + 1 */
    DWORD   fatbase;        /* FAT start sector */
    DWORD   dirbase;        /* Root directory start sector (cluster# for FAT32) */
    DWORD   database;       /* Data start sector */
#if _USE_CHDIR != 0 || _USE_CURR_DIR != 0
    DWORD curr_dir;
#endif
#if !_FS_READONLY
    DWORD   last_clust;     /* Last allocated cluster */
    DWORD   free_clust;     /* Number of free clusters */
#if _USE_FSINFO
    DWORD   fsi_sector;     /* fsinfo sector */
    BYTE    fsi_flag;       /* fsinfo dirty flag (1:must be written back) */
  //BYTE    pad2;
#endif
#endif
    BYTE    fs_type;        /* FAT sub type */
    BYTE    csize;          /* Number of sectors per cluster */
#if S_MAX_SIZ > 512U
    WORD    s_size;         /* Sector size */
#endif
    BYTE    n_fats;         /* Number of FAT copies */
    BYTE    drive;          /* Physical drive number */
#if _USE_1_BUF == 0
    BUF   buf;
#endif
} FATFS;

/* Directory object structure */
typedef struct _DIR {
  //WORD    id;         /* Owner file system mount ID */
    WORD    index;      /* Current index */
    FATFS*  fs;         /* Pointer to the owner file system object */
    DWORD   sclust;     /* Start cluster */
    DWORD   clust;      /* Current cluster */
    DWORD   sect;       /* Current sector */
} DIR;


/* File object structure */
typedef struct _FIL {
  //WORD    id;             /* Owner file system mount ID */
    BYTE    flag;           /* File status flags */
    BYTE    csect;          /* Sector address in the cluster */
    FATFS*  fs;             /* Pointer to the owner file system object */
    DWORD   fptr;           /* File R/W pointer */
    DWORD   fsize;          /* File size */
    DWORD   org_clust;      /* File start cluster */
    DWORD   curr_clust;     /* Current cluster */
    DWORD   curr_sect;      /* Current sector */
#if _FS_READONLY == 0
    DWORD   dir_sect;       /* Sector containing the directory entry */
    BYTE*   dir_ptr;        /* Ponter to the directory entry in the window */
#endif
#if _USE_LESS_BUF == 0 && _USE_1_BUF == 0
    BUF   buf;              /* File R/W buffer */
#endif
} FIL;


/* File status structure */
typedef struct _FILINFO {
    DWORD fsize;            /* Size */
    WORD fdate;             /* Date */
    WORD ftime;             /* Time */
    BYTE fattrib;           /* Attribute */
    DWORD clust;            /* Start cluster */
    UCHAR fname[8+1+3+1];   /* Name (8.3 format) */
#if _USE_LFN != 0
    UCHAR* lfn;
#endif
} FILINFO;



/* Definitions corresponds to multi partition */

#if _MULTI_PARTITION != 0   /* Multiple partition cfg */

#define LD2PD(drv) (drv >> (8-_PARTITION_MASK))       /* Get physical drive# */
#define LD2PT(drv) (drv & ((1<<_PARTITION_MASK)-1))   /* Get partition# */
#define _LOGICAL_DRIVES (_DRIVES * (1<<_PARTITION_MASK))

#else                               /* Single partition cfg */

#define LD2PD(drv) (drv)            /* Physical drive# is equal to logical drive# */
#define LD2PT(drv) 0                /* Always mounts the 1st partition */
#define _LOGICAL_DRIVES _DRIVES

#endif


/* File function return code (FRESULT) */

typedef enum {
    FR_OK = 0,          /* 0 */
    FR_NOT_READY,       /* 1 */
    FR_NO_FILE,         /* 2 */
    FR_NO_PATH,         /* 3 */
    FR_INVALID_NAME,    /* 4 */
    FR_INVALID_DRIVE,   /* 5 */
    FR_DENIED,          /* 6 */
    FR_EXIST,           /* 7 */
    FR_RW_ERROR,        /* 8 */
    FR_WRITE_PROTECTED, /* 9 */
    FR_NOT_ENABLED,     /* 10 */
    FR_NO_FILESYSTEM,   /* 11 */
    FR_INVALID_OBJECT,  /* 12 */
    FR_MKFS_ABORTED,    /* 13 */
    FR_IS_DIRECTORY,    /* 13 */
    FR_IS_READONLY,     /* 14 */
    FR_DIR_NOT_EMPTY,   /* 15 */
    FR_NOT_DIRECTORY    /* 16 */
} FRESULT;



/*-----------------------------------------------------*/
/* FatFs module application interface                  */

#if _USE_DRIVE_PREFIX == 0
FRESULT f_mount (BYTE, FATFS*);                             /* Mount/Unmount a logical drive */
FRESULT f_open (FATFS*, FIL*, const UCHAR*, BYTE);          /* Open or create a file */
FRESULT f_read (FIL*, void*, UINT, UINT*);                  /* Read data from a file */
FRESULT f_write (FIL*, const void*, UINT, UINT*);           /* Write data to a file */
FRESULT f_lseek (FIL*, DWORD);                              /* Move file pointer of a file object */
FRESULT f_close (FIL*);                                     /* Close an open file object */
FRESULT f_opendir (FATFS*, DIR*, const UCHAR*);             /* Open an existing directory */
FRESULT f_readdir (DIR*, FILINFO*);                         /* Read a directory item */
FRESULT f_stat (FATFS*, const UCHAR*, FILINFO*);            /* Get file status */
FRESULT f_getfree (FATFS*, const UCHAR*, DWORD*);           /* Get number of free clusters on the drive */
FRESULT f_sync (FIL*);                                      /* Flush cached data of a writing file */
FRESULT f_unlink (FATFS*, const UCHAR*);                    /* Delete an existing file or directory */
FRESULT f_mkdir (FATFS*, const UCHAR*);                     /* Create a new directory */
FRESULT f_chmod (FATFS*, const UCHAR*, BYTE, BYTE);         /* Change file/dir attriburte */
FRESULT f_rename (FATFS*, const UCHAR*, const UCHAR*);      /* Rename/Move a file or directory */
FRESULT f_mkfs (BYTE, BYTE, WORD);                          /* Create a file system on the drive */
FRESULT f_chdir (FATFS*, const UCHAR*);                     /* Change current directory */

#else

FRESULT f_mount (BYTE, FATFS*);                             /* Mount/Unmount a logical drive */
FRESULT f_open (FIL*, const UCHAR*, BYTE);                  /* Open or create a file */
FRESULT f_read (FIL*, void*, UINT, UINT*);                  /* Read data from a file */
FRESULT f_write (FIL*, const void*, UINT, UINT*);           /* Write data to a file */
FRESULT f_lseek (FIL*, DWORD);                              /* Move file pointer of a file object */
FRESULT f_close (FIL*);                                     /* Close an open file object */
FRESULT f_opendir (DIR*, const UCHAR*);                     /* Open an existing directory */
FRESULT f_readdir (DIR*, FILINFO*);                         /* Read a directory item */
FRESULT f_stat (const UCHAR*, FILINFO*);                    /* Get file status */
FRESULT f_getfree (const UCHAR*, DWORD*, FATFS**);          /* Get number of free clusters on the drive */
FRESULT f_sync (FIL*);                                      /* Flush cached data of a writing file */
FRESULT f_unlink (const UCHAR*);                            /* Delete an existing file or directory */
FRESULT f_mkdir (const UCHAR*);                             /* Create a new directory */
FRESULT f_chmod (const UCHAR*, BYTE, BYTE);                 /* Change file/dir attriburte */
FRESULT f_rename (const UCHAR*, const UCHAR*);              /* Rename/Move a file or directory */
FRESULT f_mkfs (BYTE, BYTE, WORD);                          /* Create a file system on the drive */
FRESULT f_chdir (const UCHAR*);                             /* Change current directory */

#endif

/* Low Level functions */
FRESULT l_opendir(FATFS* fs, DWORD cluster, DIR *dirobj);   /* Open an existing directory by its start cluster */
FRESULT l_opencluster(FATFS *fs, FIL *fp, DWORD clust);     /* Open a cluster by number as a read-only file */
FRESULT l_getfree (FATFS*, const UCHAR*, DWORD*, DWORD);    /* Get number of free clusters on the drive, limited */

#if _USE_STRFUNC
#define feof(fp) ((fp)->fptr == (fp)->fsize)
#define EOF -1
int fputc (int, FIL*);                              /* Put a character to the file */
int fputs (const char*, FIL*);                      /* Put a string to the file */
int fprintf (FIL*, const char*, ...);               /* Put a formatted string to the file */
char* fgets (char*, int, FIL*);                     /* Get a string from the file */
#endif

/* User defined function to give a current time to fatfs module */

#ifdef HAVE_RTC
DWORD get_fattime (void); /* 31-25: Year(0-127 org.1980), 24-21: Month(1-12), 20-16: Day(1-31) */
                            /* 15-11: Hour(0-23), 10-5: Minute(0-59), 4-0: Second(0-29 *2) */
#else
/* Fixed time: 1982-08-31 0:00:00, same month as the introduction of the C64 */
#  define get_fattime() 0x51f0000
#endif

/* File access control and file status flags (FIL.flag) */

#define FA_READ             0x01
#define FA_OPEN_EXISTING    0x00
#if _FS_READONLY == 0
#define FA_WRITE            0x02
#define FA_CREATE_NEW       0x04
#define FA_CREATE_ALWAYS    0x08
#define FA_OPEN_ALWAYS      0x10
#define FA__WRITTEN         0x20
#define FA__DIRTY           0x40
#endif
#define FA__ERROR           0x80


/* FAT sub type (FATFS.fs_type) */

#define FS_FAT12    1
#define FS_FAT16    2
#define FS_FAT32    3


/* File attribute bits for directory entry */

#define AM_RDO  0x01    /* Read only */
#define AM_HID  0x02    /* Hidden */
#define AM_SYS  0x04    /* System */
#define AM_VOL  0x08    /* Volume label */
#define AM_LFN  0x0F    /* LFN entry */
#define AM_DIR  0x10    /* Directory */
#define AM_ARC  0x20    /* Archive */



/* Offset of FAT structure members */

#define BS_jmpBoot          0
#define BS_OEMName          3
#define BPB_BytsPerSec      11
#define BPB_SecPerClus      13
#define BPB_RsvdSecCnt      14
#define BPB_NumFATs         16
#define BPB_RootEntCnt      17
#define BPB_TotSec16        19
#define BPB_Media           21
#define BPB_FATSz16         22
#define BPB_SecPerTrk       24
#define BPB_NumHeads        26
#define BPB_HiddSec         28
#define BPB_TotSec32        32
#define BS_55AA             510

#define BS_DrvNum           36
#define BS_BootSig          38
#define BS_VolID            39
#define BS_VolLab           43
#define BS_FilSysType       54

#define BPB_FATSz32         36
#define BPB_ExtFlags        40
#define BPB_FSVer           42
#define BPB_RootClus        44
#define BPB_FSInfo          48
#define BPB_BkBootSec       50
#define BS_DrvNum32         64
#define BS_BootSig32        66
#define BS_VolID32          67
#define BS_VolLab32         71
#define BS_FilSysType32     82

#define FSI_LeadSig         0
#define FSI_StrucSig        484
#define FSI_Free_Count      488
#define FSI_Nxt_Free        492

#define MBR_Table           446

#define DIR_Name            0
#define DIR_Attr            11
#define DIR_NTres           12
#define DIR_Chksum          13
#define DIR_CrtTime         14
#define DIR_CrtDate         16
#define DIR_FstClusHI       20
#define DIR_WrtTime         22
#define DIR_WrtDate         24
#define DIR_FstClusLO       26
#define DIR_FileSize        28



/* Multi-byte word access macros  */

#if _MCU_ENDIAN == 1    /* Use word access */
#define LD_WORD(ptr)        (WORD)(*(WORD*)(BYTE*)(ptr))
#define LD_DWORD(ptr)       (DWORD)(*(DWORD*)(BYTE*)(ptr))
#define ST_WORD(ptr,val)    *(WORD*)(BYTE*)(ptr)=(WORD)(val)
#define ST_DWORD(ptr,val)   *(DWORD*)(BYTE*)(ptr)=(DWORD)(val)
#elif _MCU_ENDIAN == 2  /* Use byte-by-byte access */
#define LD_WORD(ptr)        (WORD)(((WORD)*(volatile BYTE*)((ptr)+1)<<8)|(WORD)*(volatile BYTE*)(ptr))
#define LD_DWORD(ptr)       (DWORD)(((DWORD)*(volatile BYTE*)((ptr)+3)<<24)|((DWORD)*(volatile BYTE*)((ptr)+2)<<16)|((WORD)*(volatile BYTE*)((ptr)+1)<<8)|*(volatile BYTE*)(ptr))
#define ST_WORD(ptr,val)    *(volatile BYTE*)(ptr)=(BYTE)(val); *(volatile BYTE*)((ptr)+1)=(BYTE)((WORD)(val)>>8)
#define ST_DWORD(ptr,val)   *(volatile BYTE*)(ptr)=(BYTE)(val); *(volatile BYTE*)((ptr)+1)=(BYTE)((WORD)(val)>>8); *(volatile BYTE*)((ptr)+2)=(BYTE)((DWORD)(val)>>16); *(volatile BYTE*)((ptr)+3)=(BYTE)((DWORD)(val)>>24)
#else
#error Do not forget to set _MCU_ENDIAN properly!
#endif

#define _FATFS
#endif /* _FATFS */
